ingestion-core/src/main/java/com/mozilla/telemetry/ingestion/core/transform/PubsubMessageToObjectNode.java

```diff
@@ -595,6 +642,59 @@ private String getAndCacheBqName(String name) {
       }
     }
 
+    /**
+     * Encode the given JSON object as a compact string.
+     *
+     * <p>We maintain a BigQuery persistent UDF that can parse a variety of formats, so any format
+     * used here must first be supported in:
+     * https://github.com/mozilla/bigquery-etl/blob/master/mozfun/hist/extract/udf.sql
+     *
+     * <p>Schema validation should generally ensure that all histograms are well-formed, but we
+     * perform some light sanity checking on the values and fall back to encoding the histogram
+     * as a JSON string if anything seems off.
+     */
+    private TextNode compactHistogramEncoding(JsonNode o, String fieldName) {
+      TextNode jsonEncoding = TextNode.valueOf(Json.asString(o));
+      int histogramType = o.path("histogram_type").asInt(-1);
+      if (histogramType < 0 || histogramType > 5) {
+        incrementInvalidHistogramType();
+        return jsonEncoding;
+      }
+      long sum = o.path("sum").asLong(-1);
+      if (sum < 0) {
+        incrementInvalidHistogramSum();
+        return jsonEncoding;
+      }
```

Comment:
no-op: I wish that `jsonEncoding` could be computed lazily here, but I don't think the performance gain would be worth the reduction in readability.

Reply:
I refactored `jsonEncoding` to a function `jsonHistogramEncoding` so that we can avoid computing it unless necessary.